### PR TITLE
Implement np.count_nonzero

### PR DESCRIPTION
**Pull request by guilhermeleobas**
This PR implements `np.count_nonzero`. I didn't include tests with `axis` as a tuple because `np.sum` does not support it at this time. Maybe this observation should be included on `numpysupported.rst` as well.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Thanks for the PR. This patch is currently failing to pass tests on Python 2.7 and has some flake8 issues.

RE:

>  I didn't include tests with axis as a tuple because np.sum does not support it at this time. Maybe this observation should be included on numpysupported.rst as well.

Thanks, this seems like a good approach. A doc update would be welcomed too.

**Comment by guilhermeleobas** [comment]
> Thanks, this seems like a good approach. A doc update would be welcomed too.
Should I open another PR for this?

**Comment by stuartarchibald** [comment]
> > Thanks, this seems like a good approach. A doc update would be welcomed too.
> 
> Should I open another PR for this?

Yes please, if you have time.

**Comment by guilhermeleobas** [comment]
`Linux py35_np110` and `Linux py35_np111_cov` are breaking before tests execute with the following error:

> PackagesNotFoundError: The following packages are not available from current channels:
>  - numpy=1.10

**Comment by stuartarchibald** [comment]
> `Linux py35_np110` and `Linux py35_np111_cov` are breaking before tests execute with the following error:
> 
> > PackagesNotFoundError: The following packages are not available from current channels:
> > 
> > * numpy=1.10

@guilhermeleobas noted, thanks, we'll take a look into it.

Seems like this PR is legitimately failing on Python 2.7?

**Comment by guilhermeleobas** [comment]
Hi Stuart, 

I cannot reproduce this error. I've tried to reproduce it on a Linux machine with Python 2.7.

**Comment by guilhermeleobas** [comment]
I think this is a bug with NumPy itself instead of my code.

NumPy 1.9
```python
>>> a = np.arange(12).reshape(3, 4)
>>> a
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])
>>> np.count_nonzero(a)
11
>>> np.count_nonzero(a, axis=0)
11
```

NumPy Upstream
```python
>>> a = np.arange(12).reshape(3, 4)
>>> np.count_nonzero(a)
11
>>> np.count_nonzero(a, axis=0)
array([2, 3, 3, 3])
>>>
```

Should I skip those tests if NumPy version == 1.9?

**Comment by stuartarchibald** [comment]
> I think this is a bug with NumPy itself instead of my code.
> 
> NumPy 1.9
> 
> ```python
> >>> a = np.arange(12).reshape(3, 4)
> >>> a
> array([[ 0,  1,  2,  3],
>        [ 4,  5,  6,  7],
>        [ 8,  9, 10, 11]])
> >>> np.count_nonzero(a)
> 11
> >>> np.count_nonzero(a, axis=0)
> 11
> ```
> 
> NumPy Upstream
> 
> ```python
> >>> a = np.arange(12).reshape(3, 4)
> >>> np.count_nonzero(a)
> 11
> >>> np.count_nonzero(a, axis=0)
> array([2, 3, 3, 3])
> >>>
> ```
> 
> Should I skip those tests if NumPy version == 1.9?

Thanks for looking into this. It seems like the API has changed between 1.11 and 1.12:
* NumPy 1.11 https://docs.scipy.org/doc/numpy-1.11.0/reference/generated/numpy.count_nonzero.html
* NumPy 1.12 https://docs.scipy.org/doc/numpy-1.12.0/reference/generated/numpy.count_nonzero.html

such that 1.12 accepts an `axis` kwarg but seemingly this isn't rejected explicitly prior to this but its inclusion has no effect.

```pycon
>>> import numpy as np
>>> np.__version__
'1.11.3'
>>> np.count_nonzero(np.arange(5), anything_is_accepted_but_ignored = 'even strings')
4
```

I'm of the view that Numba should check the NumPy version and reject the use of `axis` for NumPy versions  <1.12, irrespective of this being different to what NumPy does because it's likely a bug/oversight in the original implementation. Thoughts?

**Comment by guilhermeleobas** [comment]
I've changed the code to ignore the axis parameter when Numpy version < 1.12.0. I also added a warning message for this case.

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -2867,6 +2869,26 @@ def np_imag_impl(a):
 #----------------------------------------------------------------------------
 # Misc functions
 
+@overload(np.count_nonzero)
+def np_count_nonzero(arr, axis=None):
+    if not type_can_asarray(arr):
+        raise errors.TypingError("The argument to np.count_nonzero must be array-like")
+    
+    if (NumpyVersion(np.__version__) < '1.12.0') and (axis in (None, types.none)):
+        warnings.warn("axis is not supported on Numpy versions < '1.12.0'")
```
Comment (guilhermeleobas) [inline numba/targets/arraymath.py]: Should I also test this warning?

Comment (stuartarchibald) [reply numba/targets/arraymath.py]: I think compilation should just abort, else we're just replicating a NumPy bug?

Comment (guilhermeleobas) [reply numba/targets/arraymath.py]: Hum...makes sense. I will change the code to reflect this.

**Comment by guilhermeleobas** [comment]
I think the CI is stuck

**Comment by seibert** [comment]
Azure has passed, so that's fine.  We have been keeping Travis CI around for a second opinion, though it tends to glitch out much more often.

**Comment by stuartarchibald** [comment]
@guilhermeleobas are you happy for this to be reviewed now? thanks.

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -2867,6 +2868,26 @@ def np_imag_impl(a):
...
 
+@overload(np.count_nonzero)
+def np_count_nonzero(arr, axis=None):
+    if not type_can_asarray(arr):
+        raise TypingError("The argument to np.count_nonzero must be array-like")
+
+    if (NumpyVersion(np.__version__) < '1.12.0'):
+        raise TypingError("axis is not supported on Numpy versions < 1.12.0")
+
+    if (axis in (None, types.none)):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: I think this ought to use `_is_nonelike()`, which is defined in this module.

File: numba/tests/test_np_functions.py [hunk 3]
```diff
@@ -375,6 +380,38 @@ def check(x_types, x_values):
...
+
+        pyfunc = count_nonzero
+        cfunc = jit(nopython=True)(pyfunc)
+
+        for arr, axis in arrays():
+            expected = pyfunc(arr, axis)
+            got = cfunc(arr, axis)
+            self.assertPreciseEqual(expected, got)
+
+    @unittest.skipUnless(NumpyVersion(np.__version__) < '1.12.0', "Numpy bug")
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]:
```suggestion
    @unittest.skipUnless(NumpyVersion(np.__version__) < '1.12.0', "NumPy Unsupported")
```

Comment (stuartarchibald) [reply numba/tests/test_np_functions.py]: (and fix up the version check for NumPy as per the above, also line wrap to 80 chars!)

File: numba/tests/test_np_functions.py [hunk 4]
```diff
@@ -375,6 +380,38 @@ def check(x_types, x_values):
...
+    def test_count_nonzero_exception(self):
+        pyfunc = count_nonzero
+        cfunc = jit(nopython=True)(pyfunc)
+        self.disable_leak_check()
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: don't think this is needed, the exception is being raised at typing time, not from compiled code?

File: numba/tests/test_np_functions.py [hunk 5]
```diff
@@ -375,6 +380,38 @@ def check(x_types, x_values):
         x_types = [types.complex64, types.complex128]
         check(x_types, x_values)
 
+    @unittest.skipIf(NumpyVersion(np.__version__) < '1.12.0', "Numpy bug")
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]:
```suggestion
    @unittest.skipUnless(NumpyVersion(np.__version__) < '1.12.0', "NumPy Unsupported")
```

Comment (stuartarchibald) [reply numba/tests/test_np_functions.py]: (and fix up the version check for NumPy as per the above, also line wrap to 80 chars!)

File: numba/targets/arraymath.py [hunk 6]
```diff
@@ -2867,6 +2868,26 @@ def np_imag_impl(a):
...
+        def impl(arr, axis=None):
+            arr2 = np.ravel(arr)
+            return np.sum(arr2 != 0)
+    else:
+        def impl(arr, axis=None):
+            arr2 = arr.astype(np.bool_)
+            return np.sum(arr2, axis=axis)
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: pretty sure that this will fall over with optional types, but fixing this is out of scope for this PR.

**Comment by stuartarchibald** [comment]
Thanks @guilhermeleobas, unfortunately CI seems to be unhappy with one of the unit tests, seems to be NumPy version related.

File: numba/targets/arraymath.py (hunk 6 above)
Comment (stuartarchibald) [reply numba/targets/arraymath.py]: Like this:
```python
from numba import njit
import numpy as np

@njit
def foo(c):
    x = np.arange(10)
    ax = None
    if c > 2:
        ax = 1
    return np.count_nonzero(x, axis=ax)

foo(1)
```
issue is in `np.sum` with axis impl, not in your code.

**Comment by guilhermeleobas** [comment]
I think it's ready to go!

File: numba/targets/arraymath.py (hunk 6 above)
Comment (stuartarchibald) [reply numba/targets/arraymath.py]: I know that @luk-f-a is working on the `np.sum` code to add new features, and in the process we realised that there's need to rewrite a lot of it. So as to avoid conflict/duplicated work  I think an issue reminding us of the problem above is likely best for now. Thanks.

Comment (guilhermeleobas) [reply numba/targets/arraymath.py]: Will do that.
